### PR TITLE
Fix deepcopy of primitive types

### PR DESCRIPTION
**Pull request by ndrluis**
The IcebergRootModel inherits from Pydantic RootModel, which has its own implementation of deepcopy. When deepcopy runs, it calls this `__deepcopy__` method and ignores that it's a Singleton. So, my solution was to change the order of inheritance and implement a `__deepcopy__` method for singletons that returns itself.

https://github.com/pydantic/pydantic/blob/f024d03b832d1bbcbadf76184ed14d92571a71ca/pydantic/root_model.py#L108-L116

### CODE REVIEW

**Review by syun64 — APPROVED** [review]
This is really cool 🔥 - thank you for putting in this bug fix so quickly!

This is a nit: I think we could add a test to [test_metadata.py](https://github.com/apache/iceberg-python/blob/main/tests/table/test_metadata.py) to deepcopy a TableMetadata object and ensure that the output is equal to the original one with a Schema representation that contains a mix of the issue affected types. 

Theoretically this will work, but its never a bad thing to add an explicit test to demonstrate that the issue is resolved. But other than that, this looks great! Such a speedy resolution of a rather opaque issue 🚀

**Comment by ndrluis** [comment]
Thank you, @syun64. I have added the test and used an existing metadata definition to avoid duplication.

Regarding the solution, to be honest, it was not quick. It took me several weeks of learning and then stepping away from the problem for a few weeks to return with fresh ideas on how to solve it.

**Comment by syun64** [comment]
Hi @HonahX @Fokko could we ask for your help in triggering this workflow?

File: pyiceberg/utils/singleton.py
Comment (kevinjqliu) [inline pyiceberg/utils/singleton.py:51]: nit: what do you think about adding a snippet here in the comments to talk about the motivation for overriding the `__deepcopy__` function here so that future readers don't need to dig through the commit history

File: tests/table/test_metadata.py
Comment (kevinjqliu) [inline tests/table/test_metadata.py:766]: nit: instead of checking for `model_copy` behavior here, we can move this to the `test_types.py` with something like this 
https://github.com/apache/iceberg-python/pull/838/files#diff-bd467a363c9f155a2e6a7716aeca0c351089135a53d09a66138cea4f16506ebdR624-R630

**Review by kevinjqliu — COMMENTED** [review]
Thanks for addressing this 🐛!

File: tests/table/test_metadata.py
Comment (ndrluis) [reply tests/table/test_metadata.py:766]: Thank you for the review @kevinjqliu. 

The `model_copy` with `deep=True` only runs in the update_metadata function. The idea here is to simulate this behavior, but I believe it would be better to move this test to run the `update_metadata` function with `enforce_validation` set to True. This way, we will test what really happens in the real world.

https://github.com/apache/iceberg-python/blob/a6cd0cf325b87b360077bad1d79262611ea64424/pyiceberg/table/__init__.py#L1059-L1081

What do you think?

Comment (kevinjqliu) [reply tests/table/test_metadata.py:766]: sounds good! 

I think thats the only place where `model_copy(deep=True)` is used in the codebase
https://github.com/search?q=repo%3Aapache%2Ficeberg-python%20model_copy(deep%3DTrue)%20&type=code

I would also suggest creating a new table metadata example instead of `EXAMPLE_TABLE_METADATA_V2`, looks like changing `EXAMPLE_TABLE_METADATA_V2` broke a lot of tests.

**Review by kevinjqliu — APPROVED** [review]
LGTM! 

@syun64 wdyt

**Review by syun64 — APPROVED** [review]
Yep! Fantastic fix 💯

**Comment by ndrluis** [comment]
@HonahX @Fokko, can you please trigger the workflow? I have already run the tests in my fork and everything is green.
image

**Comment by Fokko** [comment]
@ndrluis Thanks for pinging me here. This approval should be gone once the first PR has been merged. Which should be soon 🚀

**Review by HonahX — APPROVED** [review]
Thanks for the great fix!

> This approval should be gone once the first PR has been merged.

@Fokko I might have lost track. Which PR are you referring to as the first one?

**Comment by Fokko** [comment]
@HonahX I was referring to approving the CI to run, which is very annoying. Now @ndrluis has his first PR in, this approval should be gone and the CI should be kicked off right away.

**Comment by Fokko** [comment]
Thanks @ndrluis for working on this, and @syun64, @kevinjqliu and @HonahX for the quick review 🚀

**Pull request merged**
